### DISCUSSION TITLE
Creality Ender 3 v2 and BLTouch v3.1 - issues with x, y axis - crashing to the printer's frame

### DISCUSSION THREAD

**Discussion by kalko555** [Jyers/Marlin#1550] [Q&A]
Hi,
after installing BLTouch v3.1 on my Creality Ender 3 v2, I am facing the issue with x and y axis. During auto bed leveling seems like printer is trying to go outside x, y position of bed and make levelling there. By doing that it is crashing to printer's frame. It is happening for x and also y axis movement. My motherboard version is 4.2.2, I've tested many different firmwares but situation is still the same and I am not able to make auto bed levelling successfully. Manual movement of x/y/z axis is working correctly.  Probe offset X and Probe offset Y are configured correctly. 

[Video of behaviour - crashing to the printer frame visible at the end of video](https://www.youtube.com/watch?v=lEs0KsVjZMk)

I am using direct connection from BLTouch to the motherboard port (no adapter used).
I am using pre-compiled firmware version: E3V2-BLTouch-3x3-v4.2.2.bin 
I've tried many different versions 5x5, also HS but still the same issue.

Could you please advise me what I can do to fix it? Is it needed to make any configuration changes in firmware and compile it? Can you advise what kind of changes? 

Thank you.

**Comment by FinlessBob** [comment]
I have the same setup and no issue.  I noticed that it probed the front row correctly (2 touches per location) but then when it went back to the 2nd row it probed the same spot rather than move left and do the other 2 locations.
Did you reset after flashing per the instruction? Only thing I can think of.

**Reply by kalko555** [reply]
Yes, I made reset after flashing firmware. Issue still persists.

**Comment by lafreniereluc** [comment]
Hi, I've got the same problem on an Ender 3 Max. I've correctly set the bed size in the configuration.. but when I auto home, I can see the printhead is farther from the middle than expected. When I try and level, the X axis crashes and the previous two points are offset too far to that side as well.

Did you figure this out?

**Reply by lafreniereluc** [reply]
Seems like I figured it out. It's because the probe offsets are set for the Ender 3 V2, not the Max which has it on the other side of the printhead. Now to figure out the correct values...

**Reply by kalko555** [reply]
I have Ender 3 V2 and did not make any modifications in config file and issue still persists as you can see in video above. Have you figured out how to modify configuration and what needs to be exactly adjusted in order to have bed levelling working correctly? Thanks

**Reply by lafreniereluc** [reply]
I'll agree with @tititopher68-dev  here, I don't think this is a configuration problem. Yours did the first three probe points correctly and in the right spots. The fourth one was good, but it did that test like 5 times? I stopped counting. Then, instead of moving to the center of the bed, it moved to the back.

Mine crashed as well, but all probe points were offset too far to the right consistently. So mine was easily fixed with the probe offset values. Good luck!

**Comment by CraftyAce** [comment]
I've got the same problem as OP. It probes spots 1-4 fine. Spots 5 and 6 get probed in spot 4 and the bottom 3 all get probed in the bottom right corner. That's why it looks like it's probing multiple times in one spot. It seems like it is having issues making the x axis move left, but right movement works fine. Still have no clue how to fix it.

**Reply by CraftyAce** [reply]
@kalko555 It's not a long term solution, but I've figured out a tiny bandaid to slap on the problem so prints can still get done fairly well. 

Using the HS 3x3 mesh bin, I home it, then do my z offset then...I let it create a mesh/level that first time, wrong probes, grinding and all. After the mesh is created and saved to the epprom, I go into manual tuning and walk through each mesh point (funny how it can do this just fine) adjusting the z offset for each.

In Cura, toss out the g29 auto level start code line and put in M420 S1 instead to load the saved mesh. It's been working fine for me so far, I just make sure to manually tune the mesh every few small prints or after every big print. (Or any time I might have fudged up the bed springs a little on accident) 

Still hoping for a solution, but this is working for now.

**Reply by domeOo** [reply]
@CraftyAce having the same problem, any update on this?  @tititopher68-dev i tried your version 2.02h but still same crash on the x-axis while doing a new mash.

**Reply by CraftyAce** [reply]
> @CraftyAce having the same problem, any update on this? @tititopher68-dev i tried your version 2.02h but still same crash on the x-axis while doing a new mash.

Sadly, as of right now, I haven't found a solution yet. I still have high hopes for finding one.

**Reply by kalko555** [reply]
> @kalko555 It's not a long term solution, but I've figured out a tiny bandaid to slap on the problem so prints can still get done fairly well.
> 
> Using the HS 3x3 mesh bin, I home it, then do my z offset then...I let it create a mesh/level that first time, wrong probes, grinding and all. After the mesh is created and saved to the epprom, I go into manual tuning and walk through each mesh point (funny how it can do this just fine) adjusting the z offset for each.
> 
> In Cura, toss out the g29 auto level start code line and put in M420 S1 instead to load the saved mesh. It's been working fine for me so far, I just make sure to manually tune the mesh every few small prints or after every big print. (Or any time I might have fudged up the bed springs a little on accident)
> 
> Still hoping for a solution, but this is working for now.

@CraftyAce thank you for workaround solution. As there is still not permanent fix I would like to use your suggested procedure. I got to the point "manual tuning and walk through each mesh point". How are you making manual level probing of BL-touch here? I do not have option to make manual probing. I am in menu (Level -> Manual tuning) and here I have only option to manually configure z-offset for current point but no possibility to make manual z level probing with BL-touch sensor. Walking through each mesh point is working fine in manual tuning also for me. Thanks for any advice.

**Comment by GhostlyCrowd** [comment]
This is a timer issue conflict with the stepper ISR and the probe timer and a whole gamut of other timer conflicts that as far as I know are fringe case timer conflict/issues that were never fixed in maple. This was reported by me (and others with various other issues related to timers) upstream on marlin. It's an issue only in the Maple HAL not STM32HAL, and the devs who looked into it decided that since they were intending to retire Maple at some point and decided putting resources into getting all the Maple boards supported by the STM32HAL  was better use of time. They are keeping around the maple HAL right now for legacy, but it will eventually go away. 

This issue also is on more boards than just the creality ones, i had the exact issues described here on a BTT SKRE3Mini. And moving to STM32HAL solved it completely.   See https://github.com/MarlinFirmware/Marlin/issues/21429  and https://github.com/MarlinFirmware/Marlin/pull/21488 

All firmware on these boards and others if based on the Maple hal suffer from lcd flickering, random no prob deploys, tool head randomly crashes to bed or axis because an end stop trigger was missed. And so forth.

**Reply by GhostlyCrowd** [reply]
> @GhostlyCrowd I'm not sure (and not even convinced at all) about Creality v422 or v427. I am in UBL on all my Ender3v2 and Voxelab Aquila printers and have never had this issue (and others issues mentioned) and I don't think I am the only one. On the other hand, the technical services of Creality have confirmed to me the factory faults concerning the 5-pin port of the BLTouch on certain motherboards leading to malfunctions of the BLTouch. And besides, this problem is not specifically related to UBL mode. As for the STM32HAL, they bring random and incomprehensible issues such as the reversal of direction of step motors for example. The Marlin v2.0.8.2 base was very stable.

An easy test would be to get users who can repeat the issue every time like i could to compile in the stm32HAL and try,  to me as it's exactly the quarks that I observed back in may. It gave me flashbacks. The issues didn't always happen, but when the timer conflicts propagated because they just so happened to overlap with that chain of stepper movements, then bam you get issues.  Timer conflicts do still exist in the Maple hal I cannot find any information on them being fixed as again the devs do not want to push resources into maple. And times by definition are time dependent and when they just so happen to overlap and conflict you get corner case issues.  That not every one has and not all the time.   

Right now they are working on timer protection for the stepping and this https://github.com/MarlinFirmware/Marlin/pull/23187 indicates that maple also suffers the same issues and once again they devs are reluctant to fix it as again the want maple to go away, as its considered depreciated.

**Reply by GhostlyCrowd** [reply]
> @GhostlyCrowd You mention the SKR BTT E3mini v1.2. you can't compare motherboards. Those of Creality are not built in the same way. I also have a CR10V2 with a BTT E3 mini v2 and have never had a Hotend or BLTouch issue crash. I even tested this firmware (base release v1.3.4) with this card and the LCD Display of the Ender3v2 (by adapting the cable) and no malfunctions noted ...

Each Hal uses the same timers based on their platform library, and like I said, timer conflict don't always happen as an overlap condition is not always present. Which is why they are hard bugs to track down.  While the creality board may have a defect, the issues with maple still exist. There are known issues with BLTouch that can happen in some use cases on maple, as well as other timer related issues.

**Reply by CraftyAce** [reply]
> 
> An easy test would be to get users who can repeat the issue every time like i could to compile in the stm32HAL and try, to me as it's exactly the quarks that I observed back in may. It gave me flashbacks. The issues didn't always happen, but when the timer conflicts propagated because they just so happened to overlap with that chain of stepper movements, then bam you get issues. Timer conflicts do still exist in the Maple hal I cannot find any information on them being fixed as again the devs do not want to push resources into maple. And times by definition are time dependent and when they just so happen to overlap and conflict you get corner case issues. That not every one has and not all the time.
> 
> Right now they are working on timer protection for the stepping and this [MarlinFirmware#23187](https://github.com/MarlinFirmware/Marlin/pull/23187) indicates that maple also suffers the same issues and once again they devs are reluctant to fix it as again the want maple to go away, as its considered depreciated.

I mean....If you need someone that has this issue replicate every time, I basically have with several verisons of firmware, Jyers, Smith3D and such. I also have nothing but free time on my hands. I just don't have an advanced knowledge on coding and I'm new to FDM printers (Previously only had experience with Resin)

**Reply by GhostlyCrowd** [reply]
> > An easy test would be to get users who can repeat the issue every time like i could to compile in the stm32HAL and try, to me as it's exactly the quarks that I observed back in may. It gave me flashbacks. The issues didn't always happen, but when the timer conflicts propagated because they just so happened to overlap with that chain of stepper movements, then bam you get issues. Timer conflicts do still exist in the Maple hal I cannot find any information on them being fixed as again the devs do not want to push resources into maple. And times by definition are time dependent and when they just so happen to overlap and conflict you get corner case issues. That not every one has and not all the time.
> > Right now they are working on timer protection for the stepping and this [MarlinFirmware#23187](https://github.com/MarlinFirmware/Marlin/pull/23187) indicates that maple also suffers the same issues and once again they devs are reluctant to fix it as again the want maple to go away, as its considered depreciated.
> 
> I mean....If you need someone that has this issue replicate every time, I basically have with several verisons of firmware, Jyers, Smith3D and such. I also have nothing but free time on my hands. I just don't have an advanced knowledge on coding and I'm new to FDM printers (Previously only had experience with Resin)

Which board and drivers do you have?

**Reply by CraftyAce** [reply]
> 
I have the 4.2.2 board on my Ender 3 v2

**Comment by kalko555** [comment]
**@tititopher68-dev**  I've tried to compile code. I downloaded source code, moved configuration.h and configuration_adv.h from directory Marlin-bugfix-JyersUI/Configuration Files/E3V2 Tempates/BLTouch-3x3 to Marlin-bugfix-JyersUI/Marlin. I changed also configuration as you described above (Marlin / SRC / HAL / STM32F1/Hal.h). Also replaced old files by new ones (provided by you).Then I compiled code with success. I put it on the sd-card but I am not getting any response during boot up. LCD display is completely blank/ no interactive. If I compiled source code with default configuration files, compiling was succesfull and also boot process - of course without levelling possibility. Did I miss something or did I do it in wrong way? Thanks.

**Reply by kalko555** [reply]
yes, without any modifications -> compiling is successful and boot up process as well. 
After HAL.h modifications, putting config files for BL touch and replacement of files you provided -> compiling is successful, but I am not able to flash firmware to the printer. Boot up process is stacked with blank screen while sd-card with compiled bin file is inserted.

**Reply by kalko555** [reply]
@tititopher68-dev yes, correct.
